ember-resources/src/util/function.ts

```diff
@@ -138,36 +139,76 @@ export function trackedFunction<Return>(...passedArgs: UseFunctionArgs<Return>)
  * State container that represents the asynchrony of a `trackedFunction`
  */
 export class State<Value> {
-  @tracked isResolved = false;
+  @tracked isSuccessful = false;
+  @tracked isError = false;
   @tracked resolvedValue?: Value;
   @tracked error?: unknown;
+  @tracked previousResolvedValue?: Value;
 
   #fn: ResourceFn<Value>;
   #hooks: Hooks;
   #initialValue: Value | undefined;
 
-  constructor(fn: ResourceFn<Value>, hooks: Hooks, initialValue?: Value) {
+  constructor(
+    fn: ResourceFn<Value>,
+    hooks: Hooks,
+    initialValue?: Value,
+    previousResolvedValue?: Value
+  ) {
     this.#fn = fn;
     this.#hooks = hooks;
     this.#initialValue = initialValue;
+    this.previousResolvedValue = previousResolvedValue;
+    this.retry();
```

Comment:
why move retry here from the the resource body, where it used to be?

I know it's functionally the same (afaict, anyway), but from a side-effect and readability perspective, I think it makes sense to not start promised-based things in constructors

Reply:
My reasoning was that `State` could be imported directly and when instantiated without `retry` it would be in an unsupported state.

I can see three solutions to this:

1. Undo this change and embrace the unsupported state.
2. Make the initial state supported by adding an `isInitial` or `hasNotStarted` flag.
3. Wrap `this.retry()` with an `if` contional controlled by a constructor argument. This would allow instantiating it without retrying only consciously, e. g. in tests.

Reply:
I'm kinda leaning towards a secret 4th option, where maybe we make `State` an interface, and not allow instantiation of the `State` class at all? thoughts? (It was probably a bug that `State` is importable and usable as more than a type anyway)

Reply:
Hmm, I like the idea of strict immutable objects, but I don't see how I can implement this. The logic of creating those immutable objects must go somewhere, and it should not be `trackedFunction` itself.

The beauty of the `State` class is that it is a single-responsibility entity that deals with asynchrony. I can see `State` spit out `ImmutableState`s, but I don't see how I can replace `State` with `ImmutableState`. 

Please explain and I'm happy to work on it.

Reply:
(Out of the first three options, I think the second one is the most reasonable.)

Reply:
Resolving, as I think we've kinda talking about the same thing in two places -- here: https://github.com/NullVoxPopuli/ember-resources/pull/683#discussion_r1038506316